### PR TITLE
Add PasswordResetController and UserTokens

### PR DESCRIPTION
**Pull request by mbramson**
Adds a few things to the default project.

- Adds a `UserToken` schema
  - `UserToken` is associated with a user, and contains the following fields:
    - `token` - The token_string that is sent to the user and then sent back.
    - `type` - for now this should be one of `password_reset` and `email_confirmation` (although this is not enforced).
- Adds a `PasswordResetController` with the following actions:
  - `:create` - for requesting a password reset token be created and emailed to the given email.
    - Email is unimplemented for now. It still handles all the logic for generating a token.
    - Always reports that an email was sent unless the request was malformed.
  - `:show` - for verifying whether a token is legitimate or not.
    - Intended to be used by the front end to determine if the user should be presented with the form to change their password.
  - `:update` - accepts a token and a new password, and if the token is valid, changes the user associated with the token's password to the new password.
    - Deletes the token upon success.
- Adds a `UserTokenService` module responsible for creating and verifying tokens which are associated with users. Contains the following methods:
  - `generate_token/2` - accepts an email and a token type, and creates a token associated with the user with the given email.
  - `verify_token/2` - accepts a token string and a token type and returns `{:ok, valid_token}` where valid_token is the token associated with that token string and token type with the `:user` association preloaded.
    - Returns `{:error, :reason}` if the token does not exist or has expired.

Currently, only password resets are implemented, although the UserToken schema and the UserTokenService module will work exactly as they are for Email Confirmations as well (and potentially other types of tokens that need to be associated with a user).

### CODE REVIEW

File: template/$PROJECT_NAME$/test/$PROJECT_NAME$/account/user_token_service_test.exs [hunk 1]
```diff
@@ -0,0 +1,50 @@
+defmodule <%= @project_name_camel_case %>.Account.UserTokenServiceTest do
+  use <%= @project_name_camel_case %>.DataCase
+
+  alias <%= @project_name_camel_case %>.Account.UserTokenService
+
+  describe "generate_token/2" do
+    test "returns a token if the email is associated with an email" do
```
Comment (Serneum) [inline template/$PROJECT_NAME$/test/$PROJECT_NAME$/account/user_token_service_test.exs]: Should this say `if the email is associated with a user`?

File: template/$PROJECT_NAME$/test/$PROJECT_NAME$/account/user_token_service_test.exs [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      user = insert(:user)
+      {:ok, token} = UserTokenService.generate_token(user.email, :email_confirmation)
+      assert token.user.id == user.id
+    end
+
+    test "returns error tuple when no user exists with that email" do
```
Comment (Serneum) [inline template/$PROJECT_NAME$/test/$PROJECT_NAME$/account/user_token_service_test.exs]: You might want `returns an error tuple` for consistency with the next test. That or update the next test to match the others in this file

File: template/$PROJECT_NAME$/test/$PROJECT_NAME$/account/user_token_service_test.exs [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      assert {:error, :token_not_found} == result
+    end
+
+    test "returns error tuple if token exists but is wrong type" do
+      user_token = insert(:user_token, %{type: "email_confirmation"})
+      result = UserTokenService.verify_token(user_token.token, :wrong_type)
+      assert {:error, :token_not_found} == result
+    end
+
+    test "returns an error tuple if token exists but has expired" do
```
Comment (Serneum) [inline template/$PROJECT_NAME$/test/$PROJECT_NAME$/account/user_token_service_test.exs]: Same comment as above

**Comment by mbramson** [comment]
@Serneum take a look again when you get the chance. Added the new password reset controller to the generated README.

**Pull request merged**